dask/blockwise.py

```diff
@@ -254,10 +290,24 @@ def __init__(
         self.output_indices = tuple(output_indices)
         self.output_blocks = output_blocks
         self.dsk = dsk
-        self.indices = tuple(
-            (name, tuple(ind) if ind is not None else ind) for name, ind in indices
-        )
+
+        # Remove `BlockwiseDep` arguments from input indices
+        # and add them to `self.io_deps`.
+        # TODO: Remove `io_deps` and handle indexable objects
+        # in `self.indices` throughout `Blockwise`.
+        self.indices = []
         self.numblocks = numblocks
+        self.io_deps = io_deps or {}
+        for dep, ind in indices:
+            if isinstance(dep, BlockwiseDep):
+                name = "blockwise-io-" + output
+                self.indices.append((name, tuple(ind) if ind is not None else ind))
+                self.io_deps[name] = dep
+                self.numblocks[name] = dep.numblocks
+            else:
+                self.indices.append((dep, tuple(ind) if ind is not None else ind))
+        self.indices = tuple(self.indices)
```

Comment:
@gjoseph92 - This PR still needs some cleanup, but it now allows the information needed to define `io_deps` via the `indices` (as you have suggested in #7513).  For now, I am just checking if each element of `indices` is a `BlockwiseDep` element, and then I'm adding that element to `self.io_deps` if/when it is.  Therefore, this PR is now explicitly requiring all "IO dependencies" to inherit from a single base class.

The goal here is to come up with a simple `BlockwiseDep` specification that will allow me to remove low-level task fusion for the DataFrame API, but also make it possible for one of us to remove/replace the `io_deps` internals at any time (to make it easier to do more array-like things).

Note that `blockwise` currently requires `BlockwiseDep` to define a `numblocks` attribute, because I didn't want to make any Blockwise-API changes here.  There is also a `shape` attribute, but it is always `None` for the  `BlockwiseDepDict` subclass used for DataFrame IO, because it is not typically possible to know the actual shape (row count) without extra IO overhead.

cc @ian-r-rose